### PR TITLE
Use IEEE-754 interchange format in serialise-double

### PR DESCRIPTION
**Pull request by tanmaysachan**
This pull request modifies the serialise_double and unserialise_double functions to use IEEE format.

https://trac.xapian.org/ticket/673

### CODE REVIEW

File: xapian-core/common/serialise-double.cc [hunk 1]
```diff
@@ -38,175 +40,51 @@
...
-	    result += negative ? char(0x8e) : char(0x0e);
-	    result += char(exp + 128);
-	} else {
-	    if (exp < -32768 || exp > 32767) {
+	int exp;
+	frexp(v, &exp);
+	--exp;
+	exp >>= 3;
+	if (exp < -32768 || exp > 32767)
+	{
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: Why have you kept this code?

File: xapian-core/common/serialise-double.cc [hunk 2]
```diff
@@ -38,175 +40,51 @@
...
-    if (n > 1) {
-	Assert(n <= 8);
-	result[0] = static_cast<unsigned char>(result[0] | ((n - 1) << 4));
-    }
-
-    return result;
+	int check_if_big_endian = 1;
+	if (*(char *) &check_if_big_endian != 1)
+	{
+		// platform is big-endian
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: We really don't want to have to check this at compile-time every time this function is called.

And there's no need to - the endianness is already available as a preprocessor check:

```
#ifdef WORDS_BIGENDIAN
```

File: xapian-core/common/serialise-double.cc (hunk 1 above)
Comment (tanmaysachan) [reply xapian-core/common/serialise-double.cc]: oh sorry, I was supposed to remove it. Will correct it.

File: xapian-core/common/serialise-double.cc [hunk 3]
```diff
@@ -38,175 +40,51 @@
...
+		uint64_t * temp_cast_pointer =
+			reinterpret_cast<uint64_t *>(&v);
+		*temp_cast_pointer = do_bswap(*temp_cast_pointer);
+		v = *reinterpret_cast<double *>(temp_cast_pointer);
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: You quote often see code performing this sort of type punning, but it violates the C/C++ strict aliasing rules, which means undefined behaviour - e.g. see:

https://blog.qt.io/blog/2011/06/10/type-punning-and-strict-aliasing/

In many cases it "works", but it's really just a timebomb in the code waiting for a smarter compiler which optimises harder and takes advantage of the situation.

You're allowed to access the bytes which make up a type via a `char*` pointer, so the safe way to do what you want is to use `memcpy()`:

```
uint64_t temp;
static_assert(sizeof(temp) == sizeof(v), "Types must be the same size");
memcpy(&temp, &v, sizeof(temp));
temp = do_bswap(temp);
return string(reinterpret_cast<const char*>(&temp), sizeof(temp));
```

Modern compilers are smart enough to spot what's really going on here, so you should get machine code which efficiently transfers the bits making up a double from a floating point register to an integer register - e.g. for GCC on x86-64: https://godbolt.org/z/oxxCVC

File: xapian-core/common/serialise-double.cc [hunk 4]
```diff
@@ -38,175 +40,51 @@
...
-	unsigned char b = static_cast<unsigned char>(exp + 7);
-	if (negative) b |= static_cast<unsigned char>(0x80);
-	result += char(b);
-    } else {
-	if (exp >= -128 && exp < 127) {
-	    result += negative ? char(0x8e) : char(0x0e);
-	    result += char(exp + 128);
-	} else {
-	    if (exp < -32768 || exp > 32767) {
+	int exp;
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: You indentation doesn't match Xapian conventions (the `xapian-check-patch` script doesn't currently detect this case unfortunately) - probably you have your editor set wrongly.  In most cases, each indentation level should be 4 columns, you should have a tab character set to move to the next multiple of 8 columns, and indents should be "tab-filled" (so a single indent is 4 spaces; double indent a tab, and in general the indent is zero or more tabs then 0 or 4 spaces).

File: xapian-core/common/serialise-double.cc [hunk 5]
```diff
@@ -38,175 +40,51 @@
...
+		uint64_t * temp_cast_pointer =
+			reinterpret_cast<uint64_t *>(&v);
+		*temp_cast_pointer = do_bswap(*temp_cast_pointer);
+		v = *reinterpret_cast<double *>(temp_cast_pointer);
+	}
+	size_t double_size = sizeof(double);
+	char result_temp[double_size];
+	memcpy(result_temp, &v, double_size);
+	string result(result_temp, double_size);
+	return result;
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: There's no need to `memcpy()` in the little-endian case - just return a string from a cast pointer to the double.

File: xapian-core/common/serialise-double.cc [hunk 6]
```diff
@@ -38,175 +40,51 @@
...
-	if (exp == 15) {
-	    if (*p == end) {
-		throw Xapian::SerialisationError("Bad encoded double: short large exponent");
-	    }
-	    exp = bigexp | (static_cast<unsigned char>(*(*p)++) << 8);
-	    exp -= 32768;
-	} else {
-	    exp = bigexp - 128;
+	if (end - *p < 8)
+	{
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: The `{` should be on the same line as the `if`.

File: xapian-core/common/wordaccess.h [hunk 7]
```diff
@@ -27,6 +27,7 @@
 
 #include <cstdint>
 #include <type_traits>
+#include <cstring>
```
Comment (ojwb) [inline xapian-core/common/wordaccess.h]: This file does seem to use `memcpy()` without including the relevant header, but better not to make that unrelated fix in the same commit as changing the double serialisation.

Comment (tanmaysachan) [reply xapian-core/common/wordaccess.h]: Separated into 2 commits.

File: xapian-core/common/serialise-double.cc (hunk 4 above)
Comment (tanmaysachan) [reply xapian-core/common/serialise-double.cc]: I'm still not quite sure about the indentation, but I think the latest commit does follow what you say.

File: xapian-core/common/serialise-double.cc (hunk 3 above)
Comment (tanmaysachan) [reply xapian-core/common/serialise-double.cc]: Updated it accordingly, should be okay now. Thanks for the help!

**Comment by tanmaysachan** [comment]
@ojwb I don't get why appveyor is failing

**Comment by ojwb** [comment]
> tanmaysachan force-pushed the tanmaysachan:ticket673 branch 2 times, most recently from 8a6d0c8 to 611759a 6 days ago

> tanmaysachan force-pushed the tanmaysachan:ticket673 branch from 2b52ff9 to 439c8a0 6 days ago

In general it's really unhelpful to keep force-pushing a branch you've submitted for review (at least this is a small and fairly constrained change but it's still annoying).  The reviewer can't just look at the newer commits to see what you've changed in response to feedback - the simplest thing for them to do is just review the whole set of changes from scratch, which is more effort and likely to put them off.  Or they could try to work out which commit they last reviewed up to before history was rewritten, then generate a diff from there to the current branch head, review that, and then find the corresponding code in the github UI to add the review comments.  That's still extra work you've created for them which they wouldn't have had to do otherwise.

A good rule of thumb is don't force push a branch unless you're sure everyone collaborating with you on it is expecting it.  For example, sometimes people push work-in-progress branches mostly so there's a remote backup, but also so anyone interested can take a look, and they warn the branch "may be force-pushed without warning".  Or a reviewer might ask you to.

But in general it's better to make fixup commits [as described in our developer guide](https://xapian-developer-guide.readthedocs.io/en/latest/contributing/contributing-changes.htmlopen-a-pull-request-on-github).  These can then usually be squashed onto the commits they fixup automatically before merging.

File: xapian-core/common/serialise-double.cc (hunk 1 above)
Comment (ojwb) [reply xapian-core/common/serialise-double.cc]: Please don't mark review comments as resolved - that makes it much harder for the reviewer to check you've actually suitably addressed them.

File: xapian-core/common/serialise-double.cc [hunk 8]
```diff
@@ -38,175 +40,36 @@
...
+    uint64_t temp;
+    static_assert(sizeof(temp) == sizeof(v));
+    memcpy(&temp, &v, double_size);
+    temp = do_bswap(temp);
+    return string(reinterpret_cast<const char *>(&temp), double_size);
 #endif
-    // v is now in the range [1.0, 256.0)
-    exp >>= 3;
-    return exp;
+    return string(reinterpret_cast<const char *>(&v), double_size);
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: This really want to be in a `#else`.

Otherwise in the `WORDS_BIGENDIAN` case this line is dead code.  The compiler can spot that and not generate machine code for it, but it may also generate a warning about it.  Currently GCC doesn't, but e.g. clang does with `-Wunreachable-code-return`: "warning: 'return' will never be executed"

File: xapian-core/common/serialise-double.cc [hunk 9]
```diff
@@ -38,175 +40,36 @@
...
-    v = scalbn(v, (exp & 7) + 1);
-#else
-    v = ldexp(v, (exp & 7) + 1);
+string serialise_double(double v)
+{
+    size_t double_size = sizeof(double);
+#ifdef WORDS_BIGENDIAN
+    uint64_t temp;
+    static_assert(sizeof(temp) == sizeof(v));
+    memcpy(&temp, &v, double_size);
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: I think it's better to eliminate `double_size` - as a reviewer (and indeed as someone trying to follow the code) what I want to be able to see is that the right amount of data is being copied.  And while `double_size` suggests it holds the size of the double, I've seen enough misleadingly named variables in my time that I would want to check the definition - it's all-to-easy for someone to update the code but miss a place that should be changed.

In this version the size is clear from the line itself:

```
    memcpy(&temp, &v, sizeof(double));
```

I's also easier for the compiler to see the size - consider if the target type is too small:

```
#include <string.h>
#include <stdint.h>

uint32_t f(const char*p) {
    uint32_t r;
    memcpy(&r, p, sizeof(double));
    return r;
}

uint32_t g(const char*p) {
    const int double_size = sizeof(double);
    uint32_t r;
    memcpy(&r, p, double_size);
    return r;
}
```

Compiling that we get a warning when `sizeof(double)` is inlined but not if it comes through a variable (not even if I make it `const`):

```
$ gcc -Wall -W -c badcopy.c
badcopy.c: In function ‘f’:
badcopy.c:6:5: warning: ‘memcpy’ forming offset [5, 8] is out of the bounds [0, 4] of object ‘r’ with type ‘uint32_t’ {aka ‘unsigned int’} [-Warray-bounds]
     memcpy(&r, p, sizeof(double));
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
badcopy.c:5:14: note: ‘r’ declared here
     uint32_t r;
              ^
```

That's a potential stack smashing vulnerability.

File: xapian-core/common/serialise-double.cc [hunk 10]
```diff
@@ -38,175 +40,36 @@
...
-	v *= 256.0;
-    } while (v != 0.0 && --maxbytes);
-
-    n = result.size() - n;
-    if (n > 1) {
-	Assert(n <= 8);
-	result[0] = static_cast<unsigned char>(result[0] | ((n - 1) << 4));
+    if (end - *p < 8) {
+		throw Xapian::SerialisationError(
+				"Bad encoded double: insufficient data");
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: This block is over-indented.

File: xapian-core/common/serialise-double.cc (hunk 4 above)
Comment (ojwb) [reply xapian-core/common/serialise-double.cc]: There are still some problems, but I'll flag them individually.

File: xapian-core/common/serialise-double.cc [hunk 11]
```diff
@@ -38,175 +40,36 @@
...
     }
-
+    size_t double_size = sizeof(double);
+    double result;
+    memcpy(&result, *p, double_size);
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: The above line is OK (`double_size` concerns aside) for the little-endian case, but in the big-endian case you copy the data from p into a `double` here, then below copy it from there into a `uint64_t`, then byte swap that and, then copy back to the `double`.

That first copy is an extra copy you don't need.  Maybe the compiler will be able to optimise it away, but if you eliminate it then you don't need to worry about whether it will.

And that first copy to a `double` seems icky as the value being copied is not really a bit pattern for a `double` as such - we know reversing the bytes gives a `double` value.  In practice that probably won't cause problems, but it's conceivable it could be some sort of trapping value and trigger a floating point exception if the CPU loads it.  Again if we avoid this copy, that's something we can avoid having to worry about.

**Comment by ojwb** [comment]
@tanmaysachan One bit of admin - can you confirm if you're OK with the patch licensing requirements?

https://trac.xapian.org/browser/git/xapian-core/HACKING#L1392

File: xapian-core/common/serialise-double.cc (hunk 11 above)
Comment (tanmaysachan) [reply xapian-core/common/serialise-double.cc]: Ah yes, hadn't thought about the second possibility. Will fix it right away.

**Comment by tanmaysachan** [comment]
> @tanmaysachan One bit of admin - can you confirm if you're OK with the patch licensing requirements?
> 
> https://trac.xapian.org/browser/git/xapian-core/HACKING#L1392

@ojwb Yes, I'm okay with the patch licensing requirements,

File: xapian-core/common/serialise-double.cc (hunk 9 above)
Comment (tanmaysachan) [reply xapian-core/common/serialise-double.cc]: Fixed it. Thanks for the insight!

File: xapian-core/common/serialise-double.cc (hunk 8 above)
Comment (tanmaysachan) [reply xapian-core/common/serialise-double.cc]: Fixed in the latest commit.

File: xapian-core/common/serialise-double.cc (hunk 10 above)
Comment (tanmaysachan) [reply xapian-core/common/serialise-double.cc]: Fixed in the latest commit.

File: xapian-core/common/serialise-double.cc [hunk 12]
```diff
@@ -38,175 +40,153 @@
...
-    v = ldexp(v, (exp & 7) + 1);
-#endif
-    // v is now in the range [1.0, 256.0)
-    exp >>= 3;
-    return exp;
+string serialise_double(double v)
+{
+# ifdef WORDS_BIGENDIAN
+    uint64_t temp;
+    static_assert(sizeof(temp) == sizeof(v));
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: Omitting the second argument to `static_assert()` requires C++17, and currently we target C++11.  So can you add a description of what's being asserted here?

File: xapian-core/configure.ac [hunk 13]
```diff
@@ -801,6 +801,7 @@ using namespace std;
     [ac_cv_func_snprintf=unknown;break]
   )
 done
+
```
Comment (ojwb) [inline xapian-core/configure.ac]: Please avoid random whitespace changes in patches - they make review a little harder, and add noise for people trying to trace back through the history of the code.

File: xapian-core/configure.ac [hunk 14]
```diff
@@ -1216,6 +1217,31 @@ if test yes = "$enable_log"; then
...
+  AC_LANG_PROGRAM(
+    [[
+#include <cstdint>
+#include <cstring>
+using namespace std;
+    ]],
+    [[
+      double a = 2353.3523423;
+      uint64_t bit_pattern =
+      0b0100000010100010011000101011010001100110001101011011111011111111;
```
Comment (ojwb) [inline xapian-core/configure.ac]: Binary constants require C++14, we target C++11.  Use a hex constant instead.

File: xapian-core/configure.ac [hunk 15]
```diff
@@ -1216,6 +1217,31 @@ if test yes = "$enable_log"; then
...
+      if (memcmp(&bit_pattern, &a, sizeof(double)) == 0) return 0;
+      return 1;
```
Comment (ojwb) [inline xapian-core/configure.ac]: Won't this test always fail on a big endian platform?

File: xapian-core/common/serialise-double.cc [hunk 16]
```diff
@@ -38,175 +40,153 @@
...
+    double result = scalbn(mantissa, exp);
+# else
+    double result = ldexp(mantissa, exp);
+# endif
+    if (negative) result = -result;
+    return result;
 }
+
+#endif
+
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: Avoid blank lines at the end of files.

File: xapian-core/common/serialise-double.cc [hunk 17]
```diff
@@ -38,175 +40,153 @@
...
+    v -= 1.0;
+
+    for (int i = 51; i >= 0; --i) {
+	v *= 2.0;
+	if (v >= 1.0) {
+	    result |= mbp;
+	    v -= 1.0;
+	}
+	mbp >>= 1;
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: Working one bit at a time seems rather inefficient.

Can't we at least work byte-by-byte?  Multiply by 256.0, get the byte as the integer truncation of the double, then subtract the byte value from the double?  That's what the existing `serialise_double()` code does.  Here the first byte is only a partial byte (3 bits) but that's not hard to handle.

We could probably work in 4-byte chunks (multiply by 4294967296.0, get the uint32_t as the integer truncation of the double, subtract that from the double.  I think the existing code doesn't do that because it produces a variable length encoding and wants to stop when the remaining double is 0.0, but here we produce a fixed size result.

It might even be possible to just multiply by 2251799813685248.0 and cast to a uint64_t to get all the bits in one go.  We need to make sure the intermediate result is always representable as a double (remembering that double likely doesn't have IEEE properties here).

File: xapian-core/configure.ac (hunk 15 above)
Comment (ojwb) [reply xapian-core/configure.ac]: Ah no, it works so long as the integers and `double` have the same endianness.

In theory the two could differ I think, but I don't know of any platform where they actually do so slow handling for that case isn't a big deal.

File: xapian-core/configure.ac (hunk 13 above)
Comment (tanmaysachan) [reply xapian-core/configure.ac]: I pushed hurriedly, I apologize. Fixed.

File: xapian-core/common/serialise-double.cc (hunk 17 above)
Comment (tanmaysachan) [reply xapian-core/common/serialise-double.cc]: I couldn't really figure out a reason to prefer 32 bits over 26, So I modified it to use 26 bit chunks. Please correct me if I'm wrong.  
Multiplying by 2^51 does not seem like a safe option, so I left it.

Comment (tanmaysachan) [reply xapian-core/common/serialise-double.cc]: Changed the function to serialise in one go.

File: xapian-core/common/serialise-double.cc [hunk 18]
```diff
@@ -90,13 +90,23 @@ string serialise_double(double v){
...
-     * or-ing it with result, then repeating this step once
-     * more.
+     * Conversion of mantissa to bits is done by
+     * multiplying the mantissa with 2^52, converting
+     * it to a 64 bit integer representation of the original
+     * double.
      */
+
+    static_assert((uint64_t)1 << 52 <= numeric_limits<double>::max(),
+		  "Check if 2^52 can be represented by a double");
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: When you compare an integer to a `double`, first the integer gets converted to a `double`, then the two `double` values are compared (see "Conversions" section of https://en.cppreference.com/w/cpp/language/operator_arithmetic).  But we're trying to test if this integer is representable as a `double`, so we're relying on what happens if that conversion overflows.  If we'll definitely get an "overflow" or "infinity" value which doesn't compare <= the max value, all is well, but I'm struggling to work out if we can assume that's the case.

We could perhaps test based on the max exponent supported, but that's potentially complicated by what the radix is.  We currently hard require `FLT_RADIX` to be 2 in `xapian-core/api/sortable-serialise.cc`, but that requirement could be relaxed (2 seems to be ubiquitous for `double` on modern platforms, but 16 and 10 have been used on older machines).

So perhaps the best answer is to relax this check to `<` - in practice equality won't be achieved when we want the test to succeed, and by testing `<` we don't have to worry about that the conversion to `double` might give us `max()` as the overflow value if there's no infinity.

File: xapian-core/common/serialise-double.cc [hunk 19]
```diff
@@ -38,175 +40,159 @@
...
+	result = 0.0;
+	return string(reinterpret_cast<const char *>(&result),
+		      sizeof(uint64_t));
     }
 
-    int maxbytes = min(MAX_MANTISSA_BYTES, 8);
+    bool negative = (v < 0.0);
+    if (negative) {
+	v = -v;
+	result |= (uint64_t)1 << 63;
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: The coding guidelines in `HACKING` say to avoid C-style casts:

```
* Use C++ style casts (static_cast<>, reinterpret_cast<>, and const_cast<>)  
  or constructor-syntax (e.g. ``double(value)``) in preference to C style
  casts.  The syntax of the C++ casts is ugly, but they do make the
  intent much clearer which is definitely a good thing, and they avoid issues
  such as casting away const when you only meant to cast the type of a pointer.
```

So here `uint64_t(1)` (or `static_cast<uint64_t>(1)` is preferrable (and similarly in a couple of other places).

We'd like to be able to enable the GCC warning `-Wold-style-cast` which would issue a warning for such cases:

> -Wold-style-cast (C++ and Objective-C++ only)
>
>    Warn if an old-style (C-style) cast to a non-void type is used within a C++ program. The new-style casts (dynamic_cast, static_cast, reinterpret_cast, and const_cast) are less vulnerable to unintended effects and much easier to search for.

https://gcc.gnu.org/onlinedocs/gcc/C_002b_002b-Dialect-Options.html

However, currently we've had to disable this warning due to it triggering for macros in system header files on some platforms, but at some point hopefully we can turn it on again.

File: xapian-core/common/serialise-double.cc [hunk 20]
```diff
@@ -104,31 +103,38 @@ string serialise_double(double v){
 
     int exp;
     v = frexp(v, &exp);
-    v *= 2.0;
 
     if (exp == 0 && v == 0.0) {
-	result = 0;
+	result = 0.0;
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: Why this change?  `result` is a `uint64_t` so `0` is a clearer zero value for it than `0.0`.

Comment (tanmaysachan) [reply xapian-core/common/serialise-double.cc]: Sorry for the unwanted change. In the latest commit, I removed this condition altogether because I was accidentally performing this check twice. I now kept it to check at the start if V equals 0.0, and return 0 if it does.

File: xapian-core/common/serialise-double.cc (hunk 19 above)
Comment (tanmaysachan) [reply xapian-core/common/serialise-double.cc]: Fixed accordingly.

File: xapian-core/common/serialise-double.cc (hunk 18 above)
Comment (tanmaysachan) [reply xapian-core/common/serialise-double.cc]: Fixed accordingly.

File: xapian-core/common/serialise-double.cc (hunk 20 above)
Comment (ojwb) [reply xapian-core/common/serialise-double.cc]: Oh yes, this can only happen when v is 0.  Well spotted.

File: xapian-core/common/serialise-double.cc [hunk 21]
```diff
@@ -38,175 +40,152 @@
...
+    while (byte_counter >= 0) {
+	unsigned char cur_byte = *(*p + byte_counter);
+	--byte_counter;
+	int cur_byte_counter = 1 << 7;
+	for (int i = 0; i < 8; ++i) {
+	    if (cur_byte & cur_byte_counter) mantissa += pows2;
+	    pows2 /= 2.0;
+	    cur_byte_counter >>= 1;
 	}
-#endif
```
Comment (ojwb) [inline xapian-core/common/serialise-double.cc]: Again this is processing one bit at a time, which seems complex and inefficient.

Can we unserialise by doing the reverse the steps we use to serialise?

Comment (tanmaysachan) [reply xapian-core/common/serialise-double.cc]: Fixed it in the latest commit.

**Comment by ojwb** [comment]
Thanks, merged.

Your "FIXUP" commits aren't valid fixup commits that can be automatically squashed so I've had to leave them as-is.  See the documentation for `git rebase --autosquash` and `git commit --fixup` for how to make proper fixup commits.
